File: HaglofFileImporter.py
import xml.etree.ElementTree as ElementTree
import math

class HaglofXMLFileReader():

    def __init__(self):
        self.root = None

    def importXMLFile(self, xmlfile, rootTag=""):
        xml = ElementTree.parse(xmlfile)
        self.root = xml.getroot()

        # set data root, if given
        self.dataRoot = None
        if rootTag:
            self.dataRoot = self.root.find(rootTag)

        return True

    def setDataRoot(self, rootTag=""):
        self.dataRoot = self.root.find(rootTag)

        return True

    def readDiameterBySpecies(self, species):
        diameters = []

        for tree in self.dataRoot:
            if str(tree.findtext("SpeciesText")) == species:
                d = int(tree.findtext("Diameter")) / 10
                diameters.append(math.floor(d))

        return diameters

    def readAllTreeSpecies(self):
        species = []

        for tree in self.dataRoot:
            s = str(tree.findtext("SpeciesText"))
            if s not in species:
                species.append(s)
        
        return species

File: HaglofFileImporter.cpp
// Copyright (C) 2022 Tobias Helfenstein <tobias@die-softwarezimmerei.de>.
// Licensed under the GPLv3 License. See LICENSE file in the project root for license information.

#include "HaglofFileImporter.h"


void HaglofFileImporter::open(const char *filename)
{
	m_xmlfile = new QFile(filename);
	m_xmlfile->open(QFile::ReadOnly);

	this->xml.setDevice(m_xmlfile);
	this->readAll();

	// close the file after reading
	m_xmlfile->close();
}

void HaglofFileImporter::close()
{
	// check if file is open before close it
	if (m_xmlfile->isOpen())
	{
		m_xmlfile->close();
	}
}

bool HaglofFileImporter::readAll()
{
	MeasuredData entry;

	if (this->xml.readNextStartElement())
	{
		if (QString("mdiidata").compare(this->xml.name(), Qt::CaseInsensitive) == 0)
		{
			this->readHaglofData();
		}
		else
		{
			this->xml.raiseError("Keine Kluppendatei von HAGLÖF!");
		}
	}

	return true;
}

void HaglofFileImporter::readHaglofData()
{
	while (this->xml.readNextStartElement())
	{
		if (QString("data").compare(this->xml.name(), Qt::CaseInsensitive) == 0)
		{
			// create current measuring
			// HAGLOF always "0"
			this->changeMeasuring("haglof");

			// now begin reading tree entries
			this->readHgData();
		}
		else
			this->xml.skipCurrentElement();
	}
}

void HaglofFileImporter::readHgData()
{
	while (this->xml.readNextStartElement())
	{
		if (this->xml.name() == QString("Tree"))
			this->readTree();
		else
			this->xml.skipCurrentElement();
	}
}

void HaglofFileImporter::readTree()
{
	QString result;
	MeasuredData entry;

	while (this->xml.readNextStartElement())
	{
		result = "";

		if (this->xml.name() == QString("SpeciesText"))
		{
			result = this->xml.readElementText();
			entry.species = result.trimmed();
		}
		else if (this->xml.name() == QString("Diameter"))
		{
			result = this->xml.readElementText();
			entry.diameter = floor(result.trimmed().toInt()/ 10);
		}
		else if (this->xml.name() == QString("Height1"))
		{
			result = this->xml.readElementText();
			entry.length = result.trimmed().toFloat() / 10;
		}
		else
		{
			this->xml.skipCurrentElement();
		}

	}

	this->hasMeasured(entry);
}
